krabs/krabs/trace.hpp

```diff
@@ -440,7 +447,30 @@ namespace krabs {
     template <typename T>
     void trace<T>::set_default_event_callback(c_provider_callback callback)
     {
+        // C function pointers don't interact well with std::ref, so we
+        // overload to take care of this scenario.
         default_callback_ = callback;
     }
 
+    template <typename T>
+    template <typename U>
+    void trace<T>::set_default_event_callback(U& callback)
```

Comment:
Do you have an example of what you're trying to solve here? I don't think you should need these at all.
If there is a good reason to have this, you should use a single function:

```cpp
    template <typename T>
    template <typename U>
    void trace<T>::set_default_event_callback(U&& callback)
```

This can bind to all flavors of parameters correctly.

Reply:
> Do you have an example of what you're trying to solve here?
```
auto fun = [&](const EVENT_RECORD& record) {  };
krabs::user_trace trace;
trace.set_default_event_callback(fun);

```
The same approach is used in `event_filter::add_on_error_callback` or `base_provider<T>::add_on_event_callback`